### PR TITLE
FIR IDE: Add quickfix to change function return type

### PR DESCRIPTION
**Pull request by tgeng**
Add quickfix to change type for the following diagnostics
* RETURN_TYPE_MISMATCH
* COMPONENT_FUNCTION_RETURN_TYPE_MISMATCH

### CODE REVIEW

**Review by punzki — APPROVED** [review]
Minor comment

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt [hunk 1]
```diff
@@ -377,6 +377,7 @@ object DIAGNOSTICS_LIST : DiagnosticList() {
         val RETURN_TYPE_MISMATCH by error<KtExpression>(PositioningStrategy.WHOLE_ELEMENT) {
             parameter<ConeKotlinType>("expected")
             parameter<ConeKotlinType>("actual")
+            parameter<FirDeclaration>("targetDeclaration")
```
Comment (jsjeon) [inline compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:380]: Can this be more specific, e.g., `FirSimpleFunction`?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt [hunk 2]
```diff
@@ -105,4 +108,6 @@ object FirDiagnosticRenderers {
             "$list $branches or 'else' branch instead"
         }
     }
+
+    val NOT_RENDERED = Renderer<Any?> { "" }
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt:112]: I feel like I saw this before...? or maybe some other PR which isn't merged?

File: idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt [hunk 3]
```diff
@@ -74,8 +73,8 @@ abstract class ChangeCallableReturnTypeFix(
         val element = element!!
         if (element.name == null) return null
         val container = element.unsafeResolveToDescriptor().containingDeclaration as? ClassDescriptor
-        val containerName = container?.name?.takeUnless { it.isSpecial }
-        return StringPresentation.functionOrConstructorParameterPresentation(element, containerName)
+        val containerName = container?.name?.takeUnless { it.isSpecial }?.asString()
```
Comment (jsjeon) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt:76]: Nit: I saw you added something like `identifierIfNotSpecial`?

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt (hunk 1 above)
Comment (tgeng) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:380]: Sure sounds good!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt (hunk 2 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt:112]: Yep, it's in #4338 and it's not merged yet.

File: idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt (hunk 3 above)
Comment (tgeng) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt:76]: Yeah, that's also in #4338.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt (hunk 2 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDiagnosticRenderers.kt:112]: oh... so my memory is half correct. 🤦

File: idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt (hunk 3 above)
Comment (jsjeon) [reply idea/src/org/jetbrains/kotlin/idea/quickfix/ChangeCallableReturnTypeFix.kt:76]: Ditto. 🤦

**Review by jsjeon — APPROVED** [review]
Nit: in the commit message and the PR title, I think you meant "FIR", not "FIr" :)

and... one last dumb question: from my perspective, test changes in this PR are adding `/* IGNORE_FIR */` (what?!) I assumed that, according to `generated` tests, many are passing with quickfix addition, but only some are not handled yet (and thus ignored), right? Any explanation?

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ChangeTypeQuickFixFactories.kt [hunk 4]
```diff
@@ -39,29 +47,56 @@ object ChangeTypeQuickFix {
...
-        else -> null
+    }
+
+    private val KtCallableDeclaration.presentationForQuickfix: String?
+        get() {
+            val containerName = parentOfType<KtNamedDeclaration>()?.nameAsName?.takeUnless { it.isSpecial }
+            return ChangeTypeFixUtils.functionOrConstructorParameterPresentation(this, containerName?.asString())
+        }
+
+    enum class TargetType {
```
[the remark this answers is not in the document]
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ChangeTypeQuickFixFactories.kt:90]: Sounds good!

**Comment by tgeng** [comment]
Yep, your understanding is correct. I enabled more tests in `GeneratedTests` and disabled some of these newly included tests because we don't have the needed diagnostics implemented yet.

**Review by darthorimar — APPROVED** [review]

**Comment by darthorimar** [comment]
LGTM, will merge manually

**Pull request closed without merging**